node/core/proposer/src/lib.rs

```diff
@@ -0,0 +1,145 @@
+			client,
+			overseer,
+		}
+	}
+}
+
+impl<TxPool, Backend, Client> sp_consensus::Environment<Block>
+	for ProposerFactory<TxPool, Backend, Client>
+where
+	TxPool: 'static + TransactionPool<Block = Block>,
+	Client: 'static
+		+ BlockBuilderProvider<Backend, Block, Client>
+		+ ProvideRuntimeApi<Block>
+		+ HeaderBackend<Block>
+		+ Send
+		+ Sync,
+	Client::Api:
+		ParachainHost<Block> + BlockBuilderApi<Block> + ApiExt<Block, Error = sp_blockchain::Error>,
+	Backend:
+		'static + sc_client_api::Backend<Block, State = sp_api::StateBackendFor<Client, Block>>,
+	// Rust bug: https://github.com/rust-lang/rust/issues/24159
+	sp_api::StateBackendFor<Client, Block>: sp_api::StateBackend<HashFor<Block>> + Send,
+{
+	type CreateProposer = Pin<Box<
+		dyn Future<Output = Result<Self::Proposer, Self::Error>> + Send + 'static
+	>>;
+	type Proposer = Proposer<TxPool, Backend, Client>;
+	type Error = sp_blockchain::Error;
+
+	fn init(&mut self, parent_header: &Header) -> Self::CreateProposer {
+		unimplemented!()
+	}
+}
+
+pub struct Proposer<TxPool: TransactionPool<Block = Block>, Backend, Client> {
+	inner: sc_basic_authorship::Proposer<Backend, Block, Client, TxPool>,
+	client: Arc<Client>,
+	overseer: OverseerHandler,
+}
+
+impl<TxPool, Backend, Client> sp_consensus::Proposer<Block> for Proposer<TxPool, Backend, Client>
+where
+	TxPool: 'static + TransactionPool<Block = Block>,
+	Client: 'static
+		+ BlockBuilderProvider<Backend, Block, Client>
+		+ ProvideRuntimeApi<Block>
+		+ HeaderBackend<Block>
+		+ Send
+		+ Sync,
+	Client::Api:
+		ParachainHost<Block> + BlockBuilderApi<Block> + ApiExt<Block, Error = sp_blockchain::Error>,
+	Backend:
+		'static + sc_client_api::Backend<Block, State = sp_api::StateBackendFor<Client, Block>>,
+	// Rust bug: https://github.com/rust-lang/rust/issues/24159
+	sp_api::StateBackendFor<Client, Block>: sp_api::StateBackend<HashFor<Block>> + Send,
+{
+	type Transaction = sc_client_api::TransactionFor<Backend, Block>;
+	type Proposal = Pin<Box<dyn Future<Output = Result<Proposal<Block, sp_api::TransactionFor<Client, Block>>, Error>> + Send>>;
+	type Error = Error;
+
+	fn propose(
+		self,
+		mut inherent_data: InherentData,
+		inherent_digests: DigestFor<Block>,
+		max_duration: time::Duration,
+		record_proof: RecordProof,
+	) -> Self::Proposal {
+		// REVIEW: per the guide, block authors must re-send a new request for block authorship data
+		// for each block. I'm assuming that this function gets called once per block, but it is not
+		// obvious from the documentation that that assumption is in fact true.
+		let mut bitfields = Vec::new();
+		let mut candidates = Vec::new();
+
+		async move {
+			let (sender, receiver) = futures::channel::mpsc::channel(1);
+			self.overseer.send_msg(AllMessages::Provisioner(ProvisionerMessage::RequestBlockAuthorshipData(
+				// REVIEW: is this in fact the hash we want to use here?
+				self.client.info().best_hash,
+				sender,
+			)));
+			receiver.for_each_concurrent(None, |item| {
```

Comment:
This currently produces this error message:

```
    Checking polkadot-node-core-proposer v0.1.0 (/home/coriolinus/Documents/Projects/paritytech/polkadot/node/core/proposer)
error[E0277]: the trait bound `(): core::future::future::Future` is not satisfied
   --> node/core/proposer/src/lib.rs:109:13
    |
109 |             receiver.for_each_concurrent(None, |item| {
    |                      ^^^^^^^^^^^^^^^^^^^ the trait `core::future::future::Future` is not implemented for `()`

error: aborting due to previous error
```

I have no idea why that would be; it seems clear to me that `receiver` has type `Receiver`, not `()`.

Reply:
This is probably complaining about the return type of the closure and not the type of the `Receiver`.

Reply:
Intuitively this code does produce a borrowck error, btw, because it is altering `bitfields` and `candidates` concurrently.

I also think it would be better for the provisioner to handle sorting through the bitfields and candidates internally. The provisioner as described in the guide doesn't do that, but I don't see how that logically makes sense. The candidate backing system might give the provisioner several different conflicting candidates, and the provisioner should be responsible for choosing a set of a single candidate per para.

Also, some candidates may only be valid with the perspective of a certain set of bitfields. So it should be up to the provisioner to choose the bitfields and candidates together.

Last, this piece of code is set up in a way that if the provisioner never returns anything, this future will hang. There should be a timeout here in the block producer to ensure that doesn't happen.

Reply:
> This is probably complaining about the return type of the closure

That's surprising, compare the return type of this function

https://github.com/paritytech/polkadot/blob/ad5cbc85be8c893af13082fd25f581240cd747de/node/core/proposer/src/lib.rs#L117-L118

with the return type of an assumed-good function

https://github.com/paritytech/polkadot/blob/ad5cbc85be8c893af13082fd25f581240cd747de/validation/src/block_production.rs#L194-L200

Reply:
Oh, I get it now. Sorry, I'd misunderstood you. The closure _within_ `for_each_concurrent` needs to return a future. Thanks!